tls/s2n_security_policies.c

```diff
@@ -1262,7 +1262,16 @@ int s2n_validate_kem_preferences(const struct s2n_kem_preferences *kem_preferenc
             S2N_ERR_INVALID_SECURITY_POLICY);
     POSIX_ENSURE(S2N_IFF(kem_preferences->kem_count == 0, kem_preferences->kems == NULL),
             S2N_ERR_INVALID_SECURITY_POLICY);
-    POSIX_ENSURE(kem_preferences->tls13_kem_group_count <= S2N_SUPPORTED_KEM_GROUPS_COUNT, S2N_ERR_ARRAY_INDEX_OOB);
+    POSIX_ENSURE(kem_preferences->tls13_kem_group_count <= S2N_KEM_GROUPS_COUNT, S2N_ERR_ARRAY_INDEX_OOB);
+    uint32_t avail_count = 0;
+    for (int i = 0; i < kem_preferences->tls13_kem_group_count; i++) {
+        if (kem_preferences->tls13_kem_groups[i]->available) {
+            avail_count++;
+        }
+    }
+    uint32_t count = 0;
+    POSIX_ENSURE(s2n_result_is_ok(s2n_kem_preferences_groups_available(kem_preferences, &count)), S2N_ERR_INVALID_STATE);
+    POSIX_ENSURE(count == avail_count, S2N_ERR_INVALID_STATE);
```

Comment:
I'm not sure I understand the point of this check. Didn't you just duplicate the implementation of s2n_kem_preferences_groups_available and check that it produces the same result as the real implementation? What do you gain from that check?

Reply:
this was an attempt to honor the intent of the "sanity checks" alluded to above comment on L1260, after porting availability checks from build-time to runtime.

```
/* Basic sanity checks to assert that the count is 0 if and only if the associated list is NULL */
```

i'm happy to remove this if you think it's safe (for what it's worth, I do -- we should have faith in our CI).

Reply:
It's not that I don't think it's safe, I just don't understand what it's checking. Maybe I'm missing something and it just needs a comment?

But how it looks to me:
For `avail_count`, you loop over `kem_preferences` counting entries with `available`.
For `count` from `s2n_kem_preferences_groups_available`, you loop over `kem_preferences` counting entries with `available`.
When would avail_count != count?

Reply:
no, you're not missing anything. as far as i can tell, there's no case when `avail_count != count`. the intent was to ensure that we fail loudly if that invariant is violated by future code changes, but tests would catch that. the check is unnecessary; i'll remove it.